### PR TITLE
VM fault fastpath

### PR DESCRIPTION
**Pull request by alwin-joshy**
Implementation of the VM fault fastpath with some stubs implemented for other types of faults. Up to 62% faster for faulter -> handler direction and up to 67% faster for handler -> faulter direction. 

Includes

- Changed logic in c_traps files for AARCH64, x86_64, RISCV to redirect to fastpath.
- Newly implemented vm fault fastpath.
- Modified replyRecv fastpath to enable fastpathing of fault replies. 

Tests/Benchmarks

- seL4Test suite ran on AARCH64, AARCH32, x86, RISCV - All pass
- seL4bench suite + newly added benchmarks ran on AARCH64, AARCH32, x86. Could not run fault benchmarks on RISCV for some reason. 

EDIT: I said 62% and 67% faster but what I meant was 62% and 67% fewer cycles.

### CODE REVIEW

File: src/fastpath/fastpath.c
Comment (pingerino) [inline src/fastpath/fastpath.c:25]: nit: this declare before use is an old style from when we compiled the kernel with c90, it's not required anymore, and actually creates extra lines of proof. So new code doesn't need to use it.

File: src/fastpath/fastpath.c
Comment (pingerino) [inline src/fastpath/fastpath.c:132]: So we're only fastpathing the case where the scheduling context is donated? does that make sense in the context of a VM fault fastpath?

Comment (alwin-joshy) [reply src/fastpath/fastpath.c:132]: I think when I discussed this with Gernot and Kevin, we agreed that passive servers make sense for a lot of fault handlers. Additionally, I based a lot of this logic off the seL4_Call fastpath, which also only supports calling to passive servers at the moment. I can see that it would be ideal to be able to support active servers as fault handlers too, but I would need to have a better understanding of why this is not an option for seL4_Call and what could be done to make it work for this fastpath.

File: src/fastpath/fastpath.c
Comment (alwin-joshy) [reply src/fastpath/fastpath.c:25]: I see, I started with a declare as used style but changed it to be more similar to the call and replyrecv fastpaths. I will change it soon.

File: include/arch/arm/arch/fastpath/fastpath.h
Comment (lsf37) [inline include/arch/arm/arch/fastpath/fastpath.h:29]: these should be guarded by `EXCEPTION_FASTPATH`

(also no need for additional blank lines, one is enough)

File: src/arch/arm/c_traps.c
Comment (lsf37) [inline src/arch/arm/c_traps.c:65]: this function should still be inline -- the code should not change when `EXCEPTION_FASTPATH` is not enabled.

File: src/arch/arm/c_traps.c
Comment (lsf37) [inline src/arch/arm/c_traps.c:130]: please avoid unrelated whitespace changes

File: src/arch/arm/c_traps.c
Comment (lsf37) [inline src/arch/arm/c_traps.c:156]: please avoid unrelated whitespace changes

File: src/arch/arm/c_traps.c
Comment (lsf37) [inline src/arch/arm/c_traps.c:222]: please avoid style changes

File: src/fastpath/fastpath.c
Comment (pingerino) [reply src/fastpath/fastpath.c:132]: it doesn't make sense for seL4_Call as you need to pull in nearly the whole scheduler if you are switching SC's, as at minimum you need to reprogram the timer and check if you need to wake other tasks. There are subsets of that that could be pulled in, but you'd need to choose which scheduling scenario you want to fastpath. 

Also, the less we pull in the easier it is for verification to verify.

**Review by pingerino — COMMENTED** [review]
Have you confirmed that this behaviour is indeed a complete subset (valid traversal) of a slow path? Otherwise it's impossible to verify, as I learned when modifying fastpaths myself.

File: src/fastpath/fastpath.c
Comment (lsf37) [inline src/fastpath/fastpath.c:835]: please revert style changes, there should be zero code diff when the new fastpath is not configured.

**Review by pingerino — COMMENTED** [review]
Thanks for the comment on what you tested - I notice it doesn't mention running a VMM. sel4test doesn't really test the virtualisation paths well, so you should try running linux on top at least.

File: src/fastpath/fastpath.c
Comment (lsf37) [inline src/fastpath/fastpath.c:615]: Please use issues not comments for todo lists.

File: src/fastpath/fastpath.c
Comment (lsf37) [inline src/fastpath/fastpath.c:798]: This does not sound like a TODO you are planning to resolve. Adding an explanation here makes sense (esp on what is implicitly assumed), but it would be good to be slightly clearer that we expect the current implementation to work as is, and that more work will only be needed if we do not have `restart` set.

**Comment by lsf37** [comment]
When you say

> Could not run fault benchmarks on RISCV for some reason.

do you mean there weren't any or they were failing?

**Comment by lsf37** [comment]
The `rv32` simulation test is failing on `VMFault`, which I assume is related.

In general, what kind of testing has been done so far? Passing sel4test is good, but it does not say much about the fast path other than that it works in a few specific cases. Any thoughts an how to substantiate that the behaviour is the same as the slow path for all cases? Even if it works now, as long as there are no proofs, we need something that alerts us when the behaviour drifts from the slowpath (e.g. because there are slowpath changes an people have forgotten to update the fastpath accordingly).

In your tests, which configurations did you use? (SMP, MCS, HYP, verification/release/debug). You can switch on the hardware test suite on the PR by adding the `hw-test` label, but it will not configure the exception fastpath by default. 

Before we can merge this, we need to integrate it into CI. The main question for that is for which configurations we want the exception fastpath on. I don't think we want to double all tests (that would be easy enough to do, but it would cost a lot or test resources). It also can't be on for verification builds yet (the dependency for that still needs to be encoded in the `cmake` config). Any proposals for that?

To test the VMM with this, have a look at the setup running e.g. for [camkes-vm-examples](https://github.com/seL4/camkes-vm-examples/actions/runs/1525285140).

**Comment by alwin-joshy** [comment]
> 

Enabling the fault benchmarks on RISCV didn't do anything - the fault benchmarks were just always skipped for some reason.

**Comment by lsf37** [comment]
> Enabling the fault benchmarks on RISCV didn't do anything - the fault benchmarks were just always skipped for some reason.

Looking at he mainline sel4bench output it seems like those are not enabled at all for RISCV yet. Possibly because nobody has made them work yet. It'd be good to add them in general, but I'm not sure how much work that would be.

**Comment by lsf37** [comment]
For the DCO and gitlint test failures, you need to add a [`Signed-off-by` line](https://docs.sel4.systems/processes/contributing.html) and provide [body messages](https://docs.sel4.systems/processes/git-conventions.html) (Ok to leave out for commits you are planning to squash).

File: src/fastpath/fastpath.c
Comment (axel-h) [inline src/fastpath/fastpath.c:258]: Is this is used for 32 and 64 bit now, using `word_t` might be better.

Comment (alwin-joshy) [reply src/fastpath/fastpath.c:258]: Thanks, I've only been using RISCV64 so far and thought that would fix it for 32. The slowpath code which I based this off (handleVMFault in arch/riscv/kernel/vspace.c) seems to use uint64_t regardless of whether it is 32 or 64 bit. Any idea why this could be the case?

Comment (lsf37) [reply src/fastpath/fastpath.c:258]: It looks like this is wrong in the slowpath as well -- `read_stval` produces a `word_t`, and `seL4_Fault_VMFault_new` consumes a 32 bit value on rv32.

It's not catastrophic in this case, because `read_stval` will never produce values `seL4_Fault_VMFault_new` cannot consume, but it's better to fix the type.

Comment (axel-h) [reply src/fastpath/fastpath.c:258]: The RISC-V port has some places where the code could be cleaned up, because they seems just copy/paste to make it built somehow. And usually u64 often works for 32-bit also...

**Comment by lsf37** [comment]
(Have set the PR to `draft` as long as there is still work on making the tests pass)

File: src/fastpath/fastpath.c
Comment (yyshen) [inline src/fastpath/fastpath.c:115]: should this check be done earlier, when the handler_cap is available?

File: src/fastpath/fastpath.c
Comment (yyshen) [inline src/fastpath/fastpath.c:328]: It looks like VCPU exceptions are not included in this PR?

File: src/fastpath/fastpath.c
Comment (yyshen) [inline src/fastpath/fastpath.c:815]: Why are these special handling for these fault types needed? The default is restart=1 for all other fault types?

File: src/fastpath/fastpath.c
Comment (yyshen) [inline src/fastpath/fastpath.c:149]: one general comment is there are many arch-specific ifdefs. I wonder if we could have a better way to organise the code.

File: src/fastpath/fastpath.c
Comment (yyshen) [inline src/fastpath/fastpath.c:193]: should give 0x3ffffff a name.

File: src/fastpath/fastpath.c
Comment (alwin-joshy) [reply src/fastpath/fastpath.c:815]: They have some extra logic as the other types of exceptions may replace parts of the register set of the faulting thread. Right now, these are not necessary as we are only handling vm faults but they're more like stubs to expand upon later down the line.

File: src/fastpath/fastpath.c
Comment (alwin-joshy) [reply src/fastpath/fastpath.c:149]: Yeah it's not very pretty, I was considering putting the logic in architecture dependent functions but for now I thought it might be easier to see everything if it is all together.

File: src/fastpath/fastpath.c
Comment (alwin-joshy) [reply src/fastpath/fastpath.c:115]: It could be, but this is where the equivalent condition is checked in the seL4_Call fastpath. I don't think it would make a difference logically if it was earlier, but I thought it would be better to keep it consistent with the existing structure of other fastpath implementations.

File: src/fastpath/fastpath.c
Comment (alwin-joshy) [reply src/fastpath/fastpath.c:327]: This is correct. I thought they were in a different category to normal VM faults. Extending this fastpath to support them would require some as they are a different class of fault in seL4. Do you think they belong in this PR or as future work?

**Comment by alwin-joshy** [comment]
> The `rv32` simulation test is failing on `VMFault`, which I assume is related.
> 
> In general, what kind of testing has been done so far? Passing sel4test is good, but it does not say much about the fast path other than that it works in a few specific cases. Any thoughts an how to substantiate that the behaviour is the same as the slow path for all cases? Even if it works now, as long as there are no proofs, we need something that alerts us when the behaviour drifts from the slowpath (e.g. because there are slowpath changes an people have forgotten to update the fastpath accordingly).
> 
> In your tests, which configurations did you use? (SMP, MCS, HYP, verification/release/debug). You can switch on the hardware test suite on the PR by adding the `hw-test` label, but it will not configure the exception fastpath by default.
> 
> Before we can merge this, we need to integrate it into CI. The main question for that is for which configurations we want the exception fastpath on. I don't think we want to double all tests (that would be easy enough to do, but it would cost a lot or test resources). It also can't be on for verification builds yet (the dependency for that still needs to be encoded in the `cmake` config). Any proposals for that?
> 
> To test the VMM with this, have a look at the setup running e.g. for [camkes-vm-examples](https://github.com/seL4/camkes-vm-examples/actions/runs/1525285140).

I have run seL4test on both MCS and non-MCS for all architectures and added some new benchmarks to seL4bench , which I have also run on both MCS and non-MCS. The seL4bench ones were always done on release while the seL4test were done on debug. The updated benchmarks for vm faults can be found here https://github.com/seL4/sel4bench/pull/12. Note that unlike in seL4test, these benchmarks use seL4_replyRecv and hence go through the reply fastpath as well. I haven't done any testing for SMP. 

Regarding your first point, I'm not sure how this could be done without adding more comprehensive tests, though the existing ones do cover the different failure cases such as active vs passive fault handler, varying priorities for the server and the client. Otherwise, I don't how it can be validated besides from knowing that vm faults take much the same path as a regular seL4_Call, with the only notable differences being the setting of the tcbFault variable and the way in which the contents of the message is generated. Similarly, for the return path the notable differences are the way in which message contents are handled and how the badge and message info registers are not set. 

I have run most of the different applications from https://github.com/seL4/camkes-vm-examples and the behavior appears to be the same as in the master branch.

**Comment by alwin-joshy** [comment]
Reviving this. Have fixed the style issues raised (moving stuff to arch-dependent functions instead of ifdefs) and fixed issues with the configuration. Have also brought it back up to date with the slowpath as there were changes to aarch64HYP and aarch32HYP and the other fastpaths.  Currently implemented for aarch64, aarch32, riscv and x86_64.

**Comment by alwin-joshy** [comment]
@kent-mcleod Do you have some time to look at this?

**Comment by kent-mcleod** [comment]
> @kent-mcleod Do you have some time to look at this?

Once the signal fastpath PR is resolved I'll start looking at this one.

**Comment by alwin-joshy** [comment]
@kent-mcleod Could you take a look at this?

**Comment by lsf37** [comment]
@alwin-joshy can you please rebase and resolve conflicts first before we look into it more closely? Just want to make sure the test are running on the latest version.

File: src/arch/arm/c_traps.c
Comment (kent-mcleod) [inline src/arch/arm/c_traps.c:93]: Why are these changes manually in-lined from c_handle_vm_fault? If we're already expecting c_handle_vm_fault to be in-lined then resulting binary should already have the call in-lined.

File: src/fastpath/fastpath.c
Comment (kent-mcleod) [inline src/fastpath/fastpath.c:751]: This should probably be simplified to just:
```suggestion
        /* Note - this works as is for VM faults but will need to be changed when other faults are added. This is because
         * vm faults always restart the faulting thread upon reply but this is not necessarily the case with other types
         * of faults, which may instead become inactive. This can either be handled in the slowpath or redirected to the
         * slowpath, but either way it must be changed so that we do not try and forcefully switch to a thread which is
         * meant to be inactive. */
        word_t pc = getRestartPC(caller);
        setNextPC(caller, pc);
```

File: src/fastpath/fastpath.c
Comment (kent-mcleod) [inline src/fastpath/fastpath.c:778]: Is it possible to use the preprocessor defines to avoid the duplication of code here and conditionally place the original into an else branch?  Something like: 
```suggestion
    } else {
#else
```

Comment (kent-mcleod) [reply src/fastpath/fastpath.c:778]: And then at replace the `#endif` below with:
```
#ifdef CONFIG_EXCEPTION_FASTPATH
    }
#endif
```

**Comment by kent-mcleod** [comment]
Similarly to the Signal Fastpath, are you able to propose this PR with just the AARCH64 configuration supported, also are you able to show the comparison in fault benchmark results with and without the fault fastpath?

File: src/arch/arm/c_traps.c
Comment (alwin-joshy) [reply src/arch/arm/c_traps.c:93]: Sorry, I'm not 100% sure what you mean by this. `vm_fault_slowpath` is what I renamed `c_handle_vm_fault` to so that I could reuse it both here and in the fastpath.

Comment (alwin-joshy) [reply src/arch/arm/c_traps.c:93]: Doing it like this avoids adding an extra function that is basically identical, but it violates the convention to have the kernel be the exact same if the config option is disabled, so maybe it's better to change it

Comment (kent-mcleod) [reply src/arch/arm/c_traps.c:93]: The function bodies of `c_handle_data_fault` and `c_handle_instruction_fault` are identical apart from vm_fault_type_t. Previously, c_handle_vm_fault would be inlined into each of them.  If you make the CONFIG_EXCEPTION_FASTPATH changes in the original `c_handle_vm_fault` you get the fastpath without duplicating the C code I think

Comment (alwin-joshy) [reply src/arch/arm/c_traps.c:93]: Oh yeah, of course. That's my bad

**Comment by alwin-joshy** [comment]
I ran the benchmarks that I added to seL4bench (https://github.com/seL4/sel4bench/pull/18), which are based on the seL4test ways of testing vm faults, as well as the IPC benchmarks and obtained the following results

# Benchmarks - aarch64 (odroidc2) MCS, all measurements in cycles #

## Without the exception fastpath  ##

faulter -> fault handler:
	mean: 730.78
	std-dev: 7.44
	samples: 100

fault handler -> faulter:
	mean: 1135.27
	std-dev: 17.82
	samples: 100

normal reply fastpath (non-faulting):
	mean: 300.875
	std-dev: 4.13
	samples: 16

## With the exception fastpath ##

faulter -> fault handler
	mean: 288.26
	std-dev: 5.83
	samples: 100

fault handler -> faulter:
	mean: 334
	std-dev: 0
	samples: 100

normal reply fastpath (non-faulting):
	mean: 300
	std-dev: 0
	samples: 16

## Summary ##

faulter -> fault handler: ~2.5x faster
fault handler -> faulter: ~3.4x faster

**Comment by kent-mcleod** [comment]
Reposting this comment from the last fastpath PR: https://github.com/seL4/seL4/pull/793#issuecomment-1191915511
> Has there been any discussion about the lack of checkBudgetRestart() on this fastpath? It's not present on any of the other fastpaths either which may be an existing oversight?

File: src/fastpath/fastpath.c
Comment (kent-mcleod) [inline src/fastpath/fastpath.c:41]: This cap right check needs to do different things for MCS and for non-MCS.  On MCS the cap rights are checked when the cap is added to the TCB, so checking that the type is cap_endpoint_cap is sufficient.
For non-mcs, the checks need to be that the cap is an endpoint cap AND the cap has send rights, AND the cap has (either grant rights or grantreply rights):  
```
        cap_endpoint_cap_get_capCanSend(handlerCap) &&
        (cap_endpoint_cap_get_capCanGrant(handlerCap) ||
         cap_endpoint_cap_get_capCanGrantReply(handlerCap)))
```

File: src/fastpath/fastpath.c
Comment (kent-mcleod) [reply src/fastpath/fastpath.c:100]: While this functionality is part of the call slowpath, it's not part of the fault slowpath, so it shouldn't be part of this fastpath.  (The check at the top of this function already establishes that endpoint cap has the correct rights).

File: src/fastpath/fastpath.c
Comment (kent-mcleod) [inline src/fastpath/fastpath.c:712]: Can you expand the comment here to mention that VMFaults set thread's to ThreadState_Restart in the slowpath, but this is inlining the part of activateThread() that moves a thread from Restart to Running while updating the PC to RestartPC?

File: src/fastpath/fastpath.c
Comment (kent-mcleod) [inline src/fastpath/fastpath.c:44]: I think this is easier to read as "IF there's no send rights OR (there's no grand rights and theres no grant reply rights)":
```
                  (!cap_endpoint_cap_get_capCanSend(handler_cap) || (!cap_endpoint_cap_get_capCanGrant(handler_cap) &&
                                                                       !cap_endpoint_cap_get_capCanGrantReply(handler_cap))
```

Comment (alwin-joshy) [reply src/fastpath/fastpath.c]: Alright

File: include/arch/arm/arch/64/mode/fastpath/fastpath.h
Comment (kent-mcleod) [inline include/arch/arm/arch/64/mode/fastpath/fastpath.h:83]: I think this needs to be surrounded by `CONFIG_EXCEPTION_FASTPATH`

**Review by kent-mcleod — APPROVED** [review]
Once the preprocess tests pass and assuming the hardware tests pass, I'm happy for this to be merged.

**Review by lsf37 — APPROVED** [review]
From the verification side, this is also fine to merge. I have not tried to properly reason through equivalence, but at least I didn't spot any obvious deviations and problems and I didn't see any obvious C subset problems either.

So, good to go when the tests have passed.

**Pull request merged**
